addons/payment_payumoney/models/payumoney.py

```diff
@@ -0,0 +1,185 @@
+# FORM RELATED METHODS
+# --------------------------------------------------
+
+    @api.model
+    def _payumoney_form_get_tx_from_data(self, data):
+        """ Given a data dict coming from payu, verify it and find the related
+        transaction record. """
+        reference, pay_id, shasign = data.get('txnid'), data.get(
+            'mihpayid'), data.get('hash')
+        if not reference or not pay_id or not shasign:
+            error_msg = 'Payu: received data with missing reference (%s) or pay_id (%s) or shashign (%s)' % (
+                reference, pay_id, shasign)
+            _logger.error(error_msg)
+            raise ValidationError(error_msg)
+
+        tx = self.search([('reference', '=', reference)])
+        if not tx or len(tx) > 1:
+            error_msg = 'PayuMoney: received data for reference %s' % (
+                reference)
+            if not tx:
+                error_msg += '; no order found'
+            else:
+                error_msg += '; multiple order found'
+            _logger.error(error_msg)
+            raise ValidationError(error_msg)
+
+        #verify shasign
+        shasign_check = self.env['payment.acquirer']._payu_generate_sign(
+            tx.acquirer_id, 'out', data)
+        if shasign_check.upper() != shasign.upper():
+            error_msg = 'PayuMoney: invalid shasign, received %s, computed %s, for data %s' % (
+                shasign, shasign_check, data)
+            _logger.error(error_msg)
+            raise ValidationError(error_msg)
+        return tx
+
+    @api.model
+    def _payumoney_form_get_invalid_parameters(self, tx, data):
+        invalid_parameters = []
+
+        if tx.acquirer_reference and data.get('mihpayid') != tx.acquirer_reference:
+            invalid_parameters.append(
+                ('Transaction Id', data.get('mihpayid'), tx.acquirer_reference))
+        #check what is buyed
+        if float_compare(float(data.get('amount', '0.0')), tx.amount, 2) != 0:
+            invalid_parameters.append(
+                ('Amount', data.get('amount'), '%.2f' % tx.amount))
+
+        return invalid_parameters
+
+    @api.model
+    def _payumoney_form_validate(self, tx, data):
+        status = data.get('status')
+        if status == 'success':
+            tx.write({
+                'state': 'done',
+                'payumoney_txnid': data.get('mihpayid'),
+                'payumoney_id': data.get('payuMoneyId')
+            })
+            return True
+        elif status == 'pending':
+            tx.write({
+                'state': 'pending',
+                'payumoney_txnid': data.get('mihpayid'),
+                'payumoney_id': data.get('payuMoneyId')
+            })
+            return True
+        elif status == 'failure':
+            tx.write({
+                'state': 'cancel',
+                'payumoney_txnid': data.get('mihpayid'),
+                'payumoney_id': data.get('payuMoneyId')
+            })
+            return True
+        else:
```

Comment:
you may refactor elif 

status = data.get('status')
transaction_status = {
          'success': { 'state': 'done',
                            'payumoney_txnid': data.get('mihpayid'),
                             .... 
           },
          'pending': {
                      'state': 'pending',
                       ... 
           }
           ....
}

vals = transaction_status.get(status, transaction_status['error'])
return tx.write(vals)